llvm/lib/MC/GOFFObjectWriter.cpp

```diff
@@ -259,6 +633,411 @@ void GOFFObjectWriter::writeHeader() {
+}
+
+void GOFFObjectWriter::writeADAandCodeSectionSymbols(
+    MCAssembler &Asm, const MCAsmLayout &Layout) {
+  // Write ESD Records for ADA Section
+  GOFFSymbol ADAED = createWSASymbol(RootSD.EsdId, true);
+  StringRef FileName = "";
+  if (!Asm.getFileNames().empty())
+    FileName = sys::path::stem((*(Asm.getFileNames().begin())).first);
+  std::pair<std::string, std::string> CsectNames = Asm.getCsectNames();
+
+  GOFFSymbol ADA;
+  if (CsectNames.second.empty()) {
+    ADA = createPRSymbol(FileName.str().append("#S"), ADAED.EsdId);
+    ADA.BindingScope = GOFF::ESD_BSC_Section;
+  } else {
+    ADA = createPRSymbol(CsectNames.second, ADAED.EsdId);
+    ADA.BindingScope = GOFF::ESD_BSC_Library;
+  }
+  ADA.Executable = GOFF::ESD_EXE_DATA;
+  ADA.NameSpace = GOFF::ESD_NS_Parts;
+  ADA.Alignment = GOFF::ESD_ALIGN_Quadword;
+
+  ADA.SectionLength = std::max(getADASectionLength(Asm, Layout), 2u);
+  writeSymbol(ADAED, Layout);
+  writeSymbol(ADA, Layout);
+
+  // Write ESD Records for Code Section
+  GOFFSymbol ED = createEDSymbol("C_CODE64", RootSD.EsdId);
+  ED.SectionLength = 0;
+  ED.Executable = GOFF::ESD_EXE_CODE;
+  ED.ForceRent = true;
+
+  for (const MCSection &MCSec : Asm) {
+    SectionKind Kind = MCSec.getKind();
+    if (Kind.isText()) {
+      if (!ED.SectionLength)
+        ED.SectionLength = Layout.getSectionAddressSize(&MCSec);
+    }
+  }
+
+  GOFFSymbol LD = createLDSymbol(RootSD.Name, ED.EsdId);
+  LD.Executable = GOFF::ESD_EXE_CODE;
+  if (RootSD.BindingScope == GOFF::ESD_BSC_Section)
+    LD.BindingScope = GOFF::ESD_BSC_Section;
+  else
+    LD.BindingScope = GOFF::ESD_BSC_Library;
+
+  LD.ADAEsdId = GOFFObjectWriter::ADAEsdId;
+
+  EntryEDEsdId = ED.EsdId;
+  CodeLDEsdId = LD.EsdId;
+  writeSymbol(ED, Layout);
+  writeSymbol(LD, Layout);
+}
+
+void GOFFObjectWriter::writeSectionSymbols(MCAssembler &Asm,
+                                           const MCAsmLayout &Layout) {
+  bool IsOSLinkageType = false;
+  for (MCSection &S : Asm) {
+    auto &Section = cast<MCSectionGOFF>(S);
+    SectionKind Kind = Section.getKind();
+
+    if (Kind.isText()) {
+      // The only text section is the code section and the relevant ESD Records
+      // are already created. All we need to do is populate the SectionMap.
+      uint32_t RootSDEsdId = RootSD.EsdId;
+
+      GOFFSection GoffSec = GOFFSection(EntryEDEsdId, CodeLDEsdId, RootSDEsdId);
+      SectionMap.insert(std::make_pair(&Section, GoffSec));
+    } else if (Section.getName().starts_with(".lsda")) {
+      GOFFSymbol SD = RootSD;
+
+      const char *WSAClassName = "C_WSA64";
+      GOFFSymbol ED = createEDSymbol(WSAClassName, SD.EsdId);
+      if (IsOSLinkageType)
+        ED.LoadBehavior = GOFF::ESDLoadingBehavior::ESD_LB_Deferred;
+      ED.BindAlgorithm = GOFF::ESD_BA_Merge;
+      ED.Executable = GOFF::ESD_EXE_DATA;
+      ED.NameSpace = GOFF::ESD_NS_Parts;
+
+      GOFFSymbol PR = createPRSymbol(Section.getName(), ED.EsdId);
```

Comment:
In the rest of the logic, "PR" symbols are created from an MCSymbol, not a MCSection.  Is this done here because we don't have a MCSymbol covering the .ldsa section?  Maybe then a better way would be create such a symbol in the exception handling code ...

Reply:
**Short Answer**
MCSections map to a tuple of (SD, ED, LD/PR). Both can be valid but there is a functional difference (at least by the binder) for whether or not LD or PR is used.

**Longer Answer**
MCSections map to a tuple of (SD, ED, LD/PR). Such a tuple maps to a "GOFF section" (aside: my understanding is that the preferred term by the binder for this is a _Class_). Note these can have the same SD symbol, or even ED symbol, you need all three to define a section. The SD symbol owns the ED symbol, and the ED symbol owns the PR/LD symbol.

The difference is that if the third item is a LD symbol, the above is called a CAT class and when it is a PR symbol, it's a MRG class. When the binder tries to bind two different compilation units, and it sees two identically named classes in both compilation units (ie. SD and ED have same names), in the case of a CAT class it will just concatenate everything. On the other hand, in the case of a MRG class it will first merge and align PR elements with the same name in both CUs before concatenating them with other PR items. 

![image](https://github.com/llvm/llvm-project/assets/22278196/be55a7b5-e05c-49bd-9d97-f6062989bb25)

`C_MYCODE/DATA` are CAT classes and `C_EXTDATA` is a MRG class.

Reply:
My question was not really about the resulting GOFF, but rather about the structure of the LLVM implementation.  Specifically, should the PR GOFF symbol be generated in writeSectionSymbols or rather in writeSymbolDefinedInModule like the other PR GOFF symbols?

That said, your comment raises a couple of other questions in my mind:
- I don't know why we would want to characterize LD as a "section" - it doesn't own any TXT records.  I'd been considering ED as the "section" for text, and PR as the "section" for static data (i.e. those GOFF symbols that own TXT).  The LD seems much more clearly equivalent to a symbol inside the text section.
- The concatenate vs. merge distinction is triggered by the BindAlgorithm flag (ESD_BA_Concatenate vs. ESD_BA_Merge) rather than the symbol type (ED/LD/PR), isn't it?

Reply:
Also, what's the granularity of an MCSection? Because EDs are allowed to have multiple LD cards if you have multiple symbols that need to be addressed from another compilation. For example, multiple functions in a single compilation can be owned by the same ED with LD symbols pointing to each of them.

Reply:
Thanks for the correction kpneal and apologies for my mistake.

> My question was not really about the resulting GOFF, but rather about the structure of the LLVM implementation. Specifically, should the PR GOFF symbol be generated in writeSectionSymbols or rather in writeSymbolDefinedInModule like the other PR GOFF symbols?

Okay, I see what you're getting at and I can see now that it's rather inconsistent. I would actually have preferred to create the other PR symbols (primarily global variables) in this function, since they actually own text and IMO they're closer to sections than symbols as a result. 

With that said, while I need to check for the other PR symbol created here (the `ppa2list` symbol you mentioned below), but I'm certain that we are also emitting symbols for the LSDAs so moving them to `writeSymbolDefinedInModules should not be a problem.